### PR TITLE
Add support for stmatrix in the unit test HopperMatmulTest/HSH_NT_128BSwizzle

### PR DESCRIPTION
**Pull request by protonu**
This demonstrates the use of stmatrix in a multi-tile hopper matmul.

### CODE REVIEW

File: tests/cpp/test_matmul.cpp [hunk 1]
```diff
@@ -3706,7 +3708,26 @@ TEST_F(HopperMatmulTest, HSH_NT_128BSwizzle) {
 
   auto tv3 = castOp(DataType::Half, tv2);
 
-  fusion.addOutput(tv3);
+  TensorView* tv4 = nullptr;
+  if (use_smem_epilogue) {
```
Comment (jacobhinkle) [inline tests/cpp/test_matmul.cpp:3712]: Nit: you can unconditionally addOutput here then if using smem epilogue use cacheBefore to create the set op

File: tests/cpp/test_matmul.cpp [hunk 2]
```diff
@@ -3706,7 +3708,26 @@ TEST_F(HopperMatmulTest, HSH_NT_128BSwizzle) {
...
+    tv4 = set(tv3);
+
+    tv3->setMemoryType(MemoryType::Shared);
+    tv4->definition()->as<LoadStoreOp>()->setOpType(
+        LoadStoreOpType::CpAsyncBulkTensorTile);
+
+    fusion.addOutput(tv4);
+
+    // We'll use stmatrix.x4 to store from reg to shared memory
+    fusion.manage("st_matrix_m_tile", 16);
```
Comment (jacobhinkle) [inline tests/cpp/test_matmul.cpp:3723]: Just to be clear, does this mean we must use a macro such that getN(macro) is a multiple of 16?

File: tests/cpp/test_matmul.cpp [hunk 3]
```diff
@@ -3775,12 +3800,28 @@ TEST_F(HopperMatmulTest, HSH_NT_128BSwizzle) {
...
+          tv->getLoopDomain());
+      tv->setLoopDomain(s.as<IterDomain*>());
+    }
+  } else {
     auto s = mma_utils::MmaSwizzler::scheduleMmaOutputAllocation(
-        tv->getLoopDomain());
-    tv->setLoopDomain(s.as<IterDomain*>());
+        tv3c->getLoopDomain());
+    tv3c->setLoopDomain(s.as<IterDomain*>());
+    tv3c->setAllocationDomain(s.as<IterDomain*>(), true);
```
Comment (jacobhinkle) [inline tests/cpp/test_matmul.cpp:3813]: Is this necessary? The loop domain was just set so shouldnt that be the default allocation domain?

File: tests/cpp/test_matmul.cpp [hunk 4]
```diff
@@ -3775,12 +3800,28 @@ TEST_F(HopperMatmulTest, HSH_NT_128BSwizzle) {
...
+
+    // We need to split and merge dimensions which have
+    // been marked parallel/vectorized. So we revert the
+    // inner-most dimension back to serial.
+    tv3->axis(-1)->parallelize(ParallelType::Serial);
```
Comment (jacobhinkle) [inline tests/cpp/test_matmul.cpp:3818]: How does tv3 look before this line? I expected it to just be block tiled and the outer dims parallelized BIDy and BiDy by the propagation from tv2.

File: tests/cpp/test_matmul.cpp (hunk 2 above)
Comment (protonu) [reply tests/cpp/test_matmul.cpp:3723]: yes -
For the m-dimension this shouldn't be a problem since it's always 64 for Hopper.
For the n-dimension, if it's not a multiple of 16, it has to a multiple of 8, and we'll have to use stmatrix.x2 where st_matrix_n_tile will be 8.

File: tests/cpp/test_matmul.cpp (hunk 4 above)
Comment (protonu) [reply tests/cpp/test_matmul.cpp:3818]: I am not sure what I was doing last night that I saw a vectorized dim there! 
There isn't one - will remove the serial type.

File: tests/cpp/test_matmul.cpp (hunk 3 above)
Comment (protonu) [reply tests/cpp/test_matmul.cpp:3813]: This is require - I don't understand this well, maybe @zasdfgbnm can tell us the exact reason.

File: csrc/device_lower/pass/index.cpp
Comment (zasdfgbnm) [inline csrc/device_lower/pass/index.cpp:1]: nit: revert unrelated change

File: tests/cpp/test_matmul.cpp (hunk 3 above)
Comment (zasdfgbnm) [reply tests/cpp/test_matmul.cpp:3803]: In the legacy indexer, the default allocation is not the loop domain, it is the `getMaybeAllocationDomain`, which is the logical domain. So the behavior isn't really well-defined in the transition of indexers... My suggestion is, if commenting this out this test still pass, then remove this. Otherwise, keep it.

**Review by zasdfgbnm — APPROVED** [review]
Could you also post some perf data here with smem epilogue on? The command I use is posted here: https://github.com/NVIDIA/Fuser/issues/3279

File: tests/cpp/test_matmul.cpp (hunk 3 above)
Comment (protonu) [reply tests/cpp/test_matmul.cpp:3803]: Without it, the test doesn't pass

**Comment by protonu** [comment]
Observations:

1. On running `getBankConflictInfo`, I see both cases (mma with and without a stmatrix) have zero bank conflicts. The codes I used was:
```
  KernelExecutor ke;
  auto launch_constraints = LaunchParams();
  ke.compile(
      &fusion,
      {inputs.first, inputs.second},
      launch_constraints,
      matmul_cparams);

  auto bank_conflict_info =
      getBankConflictInfo(ke.kernel(), launch_constraints);

  if (bank_conflict_info.empty()) {
    debug() << "===== No bank confliction =====" << std::endl;
  } else {
    debug() << "======= Bank confliction =======" << std::endl;
    for (auto info : bank_conflict_info) {
      debug() << "Expr: " << info.first->toString() << std::endl;
      auto conflict = info.second;
      if (conflict.first > 1) {
        debug() << "input conflict: " << conflict.first << " way, ";
      }
      if (conflict.second > 1) {
        debug() << "output conflict: " << conflict.second << " way";
      }
      debug() << std::endl;
    }
    debug() << "================================" << std::endl;
  }
  fusion.printKernel();
```

**Comment by protonu** [comment]
!build
